include/nbl/builtin/hlsl/emulated/vector_t.hlsl

```diff
@@ -0,0 +1,523 @@
+    {
+        this_t output;
+
+        for (uint32_t i = 0u; i < CRTP::Dimension; ++i)
+            output.setComponent(i, CRTP::getComponent(i) + other.getComponent(i));
+
+        return output;
+    }
+    NBL_CONSTEXPR_INLINE_FUNC this_t operator-(this_t other)
+    {
+        this_t output;
+
+        for (uint32_t i = 0u; i < CRTP::Dimension; ++i)
+            output.setComponent(i, CRTP::getComponent(i) - other.getComponent(i));
+
+        return output;
+    }
+    NBL_CONSTEXPR_INLINE_FUNC this_t operator*(this_t other)
+    {
+        this_t output;
+
+        for (uint32_t i = 0u; i < CRTP::Dimension; ++i)
+            output.setComponent(i, CRTP::getComponent(i) * other.getComponent(i));
+
+        return output;
+    }
+
+    DEFINE_OPERATORS_FOR_TYPE(float32_t)
+    DEFINE_OPERATORS_FOR_TYPE(float64_t)
+    DEFINE_OPERATORS_FOR_TYPE(uint16_t)
+    DEFINE_OPERATORS_FOR_TYPE(uint32_t)
+    DEFINE_OPERATORS_FOR_TYPE(uint64_t)
+    DEFINE_OPERATORS_FOR_TYPE(int16_t)
+    DEFINE_OPERATORS_FOR_TYPE(int32_t)
+    DEFINE_OPERATORS_FOR_TYPE(int64_t)
+
+    NBL_CONSTEXPR_INLINE_FUNC ComponentType calcComponentSum()
+    {
+        ComponentType sum = ComponentType::create(0);
+        for (uint32_t i = 0u; i < CRTP::Dimension; ++i)
+            sum = sum + CRTP::getComponent(i);
+
+        return sum;
+    }
+};
+
+#undef DEFINE_OPERATORS_FOR_TYPE
+
+template<typename T, uint32_t N>
+struct CRTPParentStructSelector
+{
+    using type = void;
+};
+template<typename T>
+struct CRTPParentStructSelector<T, 2>
+{
+    using type = _2_component_vec<T>;
+};
+template<typename T>
+struct CRTPParentStructSelector<T, 3>
+{
+    using type = _3_component_vec<T>;
+};
+template<typename T>
+struct CRTPParentStructSelector<T, 4>
+{
+    using type = _4_component_vec<T>;
+};
+
+}
+
+template<typename T, uint32_t N>
+using emulated_vector_t = emulated_vector_impl::emulated_vector<T, typename emulated_vector_impl::CRTPParentStructSelector<T, N>::type>;
+template<typename T>
+using emulated_vector_t2 = emulated_vector_impl::emulated_vector<T, typename emulated_vector_impl::CRTPParentStructSelector<T, 2>::type>;
+template<typename T>
+using emulated_vector_t3 = emulated_vector_impl::emulated_vector<T, typename emulated_vector_impl::CRTPParentStructSelector<T, 3>::type>;
+template<typename T>
+using emulated_vector_t4 = emulated_vector_impl::emulated_vector<T, typename emulated_vector_impl::CRTPParentStructSelector<T, 4>::type>;
+
+// TODO: better implementation
+template<typename ComponentType, typename VecType>
+struct is_valid_emulated_vector
+{
+    NBL_CONSTEXPR_STATIC bool value = is_same_v<VecType, emulated_vector_t2<ComponentType> > ||
+        is_same_v<VecType, emulated_vector_t3<ComponentType> > ||
+        is_same_v<VecType, emulated_vector_t4<ComponentType> >;
+};
+
+template<typename U, typename T, typename I = uint32_t>
+struct array_get
+{
+    T operator()(NBL_REF_ARG(U) vec, const I ix)
+    {
+        return vec[ix];
+    }
+};
```

Comment:
the base unspecialized declaration should be in `functional.hlsl`

Reply:
I need `array_get` and `array_set` to be visible in `cpp_compat/intrinsics.h` and moving it to the `functional.hlsl` header would cause circular dependency. Will move it to `hlsl/array_accessors.hlsl` instead.

Reply:
maybe move to `functional/array_accessors.hlsl` then you can include in both `intrinsics.h` and `functional.hlsl` (although this is moot, since already functional takes intrinsics)